drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,265 @@
+		last_error = atoi(&buf[strlen(AT_CMD_CMS_STR)]);
+	} else {
+		last_error = 0;
+	}
+
+	if (tmpstr) {
+		buf[tmpstr - buf] = '\0';
+	}
+
+	return return_code;
+}
+
+static void callback_worker(struct k_work *item)
+{
+	struct callback_work_item *data =
+		CONTAINER_OF(item, struct callback_work_item, work);
+
+	received_data_handler(data->data, data->data_len);
+
+	k_free(data);
+}
+
+
+static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int bytes_read;
+	char buf[RESPONSE_MAX_LEN];
+	enum at_cmd_rc return_code;
+	bool write_with_response = false;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("Socket thread started");
+
+	for (;;) {
+		bytes_read = recv(common_socket_fd, buf, sizeof(buf), 0);
+
+		if (bytes_read < 0) {
+			LOG_ERR("Socket recv failed with err %d", bytes_read);
+
+			if (bytes_read == -EAGAIN) {
+				if (open_socket() == 0) {
+					goto failed;
+				}
+			}
+
+			LOG_ERR("Unrecoverable reception error, thread killed");
+			return;
+		} else if (bytes_read == sizeof(buf) &&
+			   buf[bytes_read - 1] != '\0') {
+
+			LOG_DBG("AT message to large for reception buffer");
+			goto failed;
+		}
+
+		return_code = get_return_code(buf);
+
+		if (return_code != AT_CMD_NO_RETURN_CODE) {
+			if (response_buf_len > 0) {
+				if (response_buf_len > strlen(buf)) {
+					strncpy(response_buf,
+						buf,
+						response_buf_len);
+				} else {
+					LOG_ERR("Response buffer not large enough");
+					response_buf = NULL;
+				}
+
+				write_with_response = true;
+				response_buf_len = 0;
+			}
+
+			k_msgq_put(&return_code_msq, &return_code, K_FOREVER);
+		} else {
+			write_with_response = false;
+		}
+
+		if (strlen(buf) == 0) {
+			continue;
+		}
+
+		if (received_data_handler) {
+			if (!write_with_response) {
+				struct callback_work_item *item =
+				k_malloc(sizeof(struct callback_work_item));
+
+				if (!item) {
+					LOG_DBG("Failed to allocate work item");
+					continue;
+				}
+
+				k_work_init(&item->work, callback_worker);
+
+				item->data_len = strlen(buf);
+				strcpy(item->data, buf);
+
+				k_work_submit(&item->work);
+			}
+		}  else {
+			LOG_DBG("No handler registered for received data");
+		}
+		continue;
+
+failed:
+			return_code = AT_CMD_ERROR;
+			k_msgq_put(&return_code_msq, &return_code, K_FOREVER);
+	}
+}
+
+enum at_cmd_rc at_cmd_write(const char *const cmd)
+{
+	int bytes_sent;
+	enum at_cmd_rc return_code;
+
+	k_mutex_lock(&cmd_pending, K_FOREVER);
+
+	LOG_DBG("Sending command %s", cmd);
+	bytes_sent = send(common_socket_fd, cmd, strlen(cmd), 0);
+
+	if (bytes_sent < 0) {
+		LOG_ERR("Failed to send AT command");
+	} else {
+		k_msgq_get(&return_code_msq, &return_code, K_FOREVER);
+	}
+
+	LOG_DBG("Bytes sent: %d", bytes_sent);
+
+	k_mutex_unlock(&cmd_pending);
+
+	return return_code;
+}
+
+enum at_cmd_rc at_cmd_write_with_response(const char *const cmd,
+					  char *buf,
+					  size_t buf_len)
+{
+	enum at_cmd_rc return_code;
+
+	response_buf     = buf;
```

Comment:
This should be protected with the mutex as well, In case two thread call `at_cmd_write_with_response` at the same time. I suggest to extract `at_cmd_write`  contents to a static function (w/o mutex protection), and then call this static function from the API functions (`at_cmd_write`, `at_cmd_write_with_response`). In both cases we should protect the new function call any additional code within API function with a common mutex.

Reply:
ok